SQL Server typos and keyword completion modifier

sevin7676 opened the pull request:
This contains a few different things:

1. Fixed a few of typos in files related SQL Server mode 
2. Added a few snippets and enhanced others for SQL Server mode
3. Override `getCompletions` for SQL Server mode to ensure proper case of completions and use intelligent `meta` instead of `keyword` for all completions.

![image](https://cloud.githubusercontent.com/assets/5360920/7226982/6b545d0e-e71b-11e4-81b8-bef32af18e9e.png)


~~3. Added `completionModifier` to text_highlight_rules, which is an optional function that can be specified by a mode's highlight rules that inherit from text_highlight_rules.
4. Modified `getCompletions` in text mode to call the `completionModifier` if it exists.~~

~~The completion modifier allows modifying keyword completions, which is useful for a few things:~~
~~- Remove keywords that are obsolete from the completor, but they still need to have syntax highlighting~~
~~- Modify the `meta` display in the completer to provide more information than `keyword` for some things~~
~~- Modify the case of completions, use case:  SQL Server mode has completions that should be all caps, and others that should not (even though all keywords are case-insensitive for highlighting, its simply recommended by Microsoft to use all caps for most keywords).~~


~~I modified SQL Server mode to use the completion modifier, and I went ahead and modified Javascript mode as well.~~

~~If you don't like the completion modifier and the javascript modifications, I can remove them and simply do a one off for SQL Server mode completions by overriding `getCompletions` for SQL Server mode. I decided to make the completion modifier standard because I think it could benefit almost all modes.~~

~~Javascript mode completion popup with modified `meta` for keywords:~~

![image](https://cloud.githubusercontent.com/assets/5360920/7220046/38742772-e688-11e4-89d6-baf8b0693d62.png)

nightwing commented:
I like the result of using `completionModifier`, however having it called from `getCompletions` doesn't seem to simplify the code, that is overriding `getCompletions` would require same amount of code which will work a bit faster. Also we could use information about obsolete keywords from the original keyword map instead of throwing it out at https://github.com/ajaxorg/ace/blob/master/lib/ace/mode/text_highlight_rules.js#L220

sevin7676 commented:
Good point. I think it would be valuable to have a standardized way to have create more intelligent keyword completions for each mode, but I don't have the time to figure out the best implementation.

I removed all changes that are not for SQL Server mode and added an override for `getCompletions` for SQL Server mode only. I am caching the result of this function as it takes ~15ms to execute *(compared to ~1ms when there is no extra filtering)*. I don't see any harm in caching the result, do you?

nightwing commented:
Are you sure about caching? autocompleter doesn't modify array of matches it receives from getCompletions https://github.com/ajaxorg/ace/blob/master/lib/ace/autocomplete.js#L237.

nightwing commented on lib/ace/mode/sqlserver.js:
```diff
@@ -46,6 +46,33 @@ oop.inherits(Mode, TextMode);
+        var keywords = this.$keywordList || this.$createKeywordList();
+        var types = session.$mode.$highlightRules.keywordTypes;
+    
+        return keywords.map(function(word) {
```
could we just use types skipping keywordList entirely?
keywordList removes some information that highlighter doesn't need and here we are restoring that using slow indexOf in a large array

nightwing commented on lib/ace/mode/sqlserver_highlight_rules.js:
```diff
@@ -147,7 +141,7 @@ var SqlServerHighlightRules = function() {
             regex: "'",
```
looks like the `// ' string` rule a little [bellow](https://github.com/ajaxorg/ace/pull/2460/files#diff-de016821e1a0be41f4be155e106e0760R162) is unnecessary.

sevin7676 commented:
Good point about skipping the keyword list, the custom completions is now simpler and its only compiled once.

sevin7676 commented on lib/ace/mode/sqlserver_highlight_rules.js:
```diff
@@ -164,14 +154,11 @@ var SqlServerHighlightRules = function() {
             start: "/\\*",
             end: "\\*/"
         }, {
-            token: "string", // ' string
```
I think this was the unnecessary rule you were referring to, correct? I think I copied that rule from somewhere else

nightwing commented:
Merged.
Thanks!
